src/runtime/metal.cpp

```diff
@@ -789,6 +789,22 @@ WEAK int halide_metal_run(void *user_context,
 
     set_compute_pipeline_state(encoder, pipeline_state);
 
+    int num_kernel_args = 0;
+    for (int i = 0; args[i] != nullptr; i++) {
+        ++num_kernel_args;
+    }
+
+    // NOTE(slomp): arg_sizes[] must terminate with a zero!
```

Comment:
Why is that necessary? From a casual inspection it appears that the rest of the code here accesses the arg_sizes array by index and not null termination

Reply:
I was just trying to preserve the original semantics of `arg_sizes[]`, in case someone changes code in the future and still relies on the fact it is zero-terminated.

I'd say once we have all runtimes using `halide_type_t arg_types[]`, we should probably add an extra `num_args` to the `halide_*_run()` interface.

Reply:
I think this could be done more simply by just doing the size calculations inline and skipping the intermediate buffer entirely; it's only a couple of places. See this gist for an example:

https://gist.github.com/steven-johnson/7811391b60fd21672f36be82d4ec9a1f